### PR TITLE
Portal credential enhancements rebased

### PR DESCRIPTION
**Pull request by bradymiller**
feature by @sjpadgett 
just doing some minor work on it
Continued from:
https://github.com/openemr/openemr/pull/2715

Up For Grabs demo here: https://www.open-emr.org/wiki/index.php/Development_Demo#Gamma_-_Up_For_Grabs_Demo

### CODE REVIEW

**Comment by sjpadgett** [comment]
Hi @bradymiller Because you already have this rebased and ready for master(when finished with your changes) suggest we just merge this PR instead of doing the whole merge with my branch.
I don't care about credit as that has never been my motivation anyway. I just care that it gets in to finally make stable portal login, reset and self register. From my testing, it's never worked better, yeah.

So i'll just close my PR and you can work from this PR then bring in when you feel ready.
If I need to i'll build the patch. We can just cherry pick this back to 5.0.2 because the only difference, if I remember correctly, will be a couple translations using new alternative phrases in 5.0.3 (which is kind of clever really).

Let me know.

**Comment by bradymiller** [comment]
hi @sjpadgett , When ready, plan to rebase it all into 1 commit with you as author (then you will be blamed :) ) and I'll then cherry-pick it over to rel-502 (agree is should come over with minimal merge conflicts). I think it should be ready by tomorrow (just need to sort out what to do on the userIsUnique call(it may even be fine as is after I look at it more closely), test my changes (already had fun watching it make 1000 passwords ) ), and then 1 last security audit). thanks! -brady

**Comment by sjpadgett** [comment]
> do on the userIsUnique

Just be sure to keep the return flags consistent with what they are now. It should scan account name and login names. Don't recall if I made search binary though.(should be)

File: portal/account/account.php [hunk 1]
```diff
@@ -15,11 +15,13 @@
 require_once(dirname(__FILE__) . "/../../src/Common/Session/SessionUtil.php");
 OpenEMR\Common\Session\SessionUtil::portalSessionStart();
 
-if ($_SESSION['register'] === true && isset($_SESSION['pid'])) {
+if ($_SESSION['register'] === true && isset($_SESSION['pid']) ||
+    ($_SESSION['credentials_update'] === 1 && isset($_SESSION['pid'])) ||
+    ($_SESSION['itsme'] === 1 && isset($_SESSION['password_update']))) {
```
Comment (bradymiller) [inline portal/account/account.php]: TODO:
Working through code block now.

File: portal/get_patient_info.php [hunk 2]
```diff
@@ -63,55 +63,81 @@
...
     privStatement($sqlUpdatePwd, array(
         $new_hash,
         $new_salt,
+        $_POST['login_uname'],
         $auth[COL_ID]
     ));
 } else {
-    $tmp = oemr_password_hash($plain_code, $auth[COL_POR_SALT]);
+    $tmp = oemr_password_hash($_POST['pass'], $auth[COL_POR_SALT]);
+    if ($password_update === 2) {
```
Comment (bradymiller) [inline portal/get_patient_info.php]: will place a comment here

File: portal/index.php [hunk 3]
```diff
@@ -27,15 +27,58 @@
...
     echo xlt('Patient Portal is turned off');
     exit;
 }
-
+$auth['portal_pwd'] = '';
+if (isset($_GET['woops'])) {
+    unset($_GET['woops']);
+    unset($_SESSION['password_update']);
+}
+if (isset($_GET['forward'])) {
```
Comment (bradymiller) [inline portal/index.php]: will place some comments in this code block

File: portal/index.php [hunk 4]
```diff
@@ -27,15 +27,58 @@
...
+        OpenEMR\Common\Session\SessionUtil::portalSessionCookieDestroy();
+        die(xlt("Your one time credential reset link has expired. Reset and try again.") . "time:$validate time:" . time());
+    }
+    $_SESSION['pin'] = substr($parse, 0, 6);
+    $_SESSION['forward'] = $auth['portal_onetime'];
+    $_SESSION['portal_username'] = $auth['portal_username'];
+    $_SESSION['portal_login_username'] = '';
+    $_SESSION['password_update'] = 2;
+    $_SESSION['onetime'] = $auth['portal_pwd'];
+    sqlStatementNoLog("UPDATE `patient_access_onsite` SET `portal_onetime`=NULL WHERE BINARY `portal_onetime` = ?", [$auth['portal_onetime']]);
```
Comment (bradymiller) [inline portal/index.php]: yep, the above line was a bad idea by me :)
(I basically broke the token since the sql entry is needed...)

Comment (sjpadgett) [reply portal/index.php]: maybe just clear on new credential save in get_patient_info

File: portal/account/account.php [hunk 5]
```diff
@@ -31,6 +33,27 @@
     $_SESSION['language_choice'] = (int) $_REQUEST['value'];
     echo 'okay';
     exit();
+} elseif ($action == 'userIsUnique') {
```
Comment (bradymiller) [inline portal/account/account.php]: It looks like I can protect this function, but not allowing use for the 1st comparison above (ie. the registration thing), which will prevent unauthorized bad actors from doing brute force attacks to identify active portal_login_username. Gonna do some testing. @sjpadgett , this is basically the last thing to do here; I should have it completed and ready for commit tomorrow evening; btw, the testing on all these features have been super intuitive and awesome (ie. very nice work!).

Comment (bradymiller) [reply portal/account/account.php]: for me:
this comparison will cover the script call from portal/account/index_reset.php:
isset($_SESSION['pid']) && isset($_SESSION['patient_portal_onsite_two'])

File: portal/get_patient_info.php [hunk 6]
```diff
@@ -63,55 +63,84 @@
...
+// 2 is flag for one time credential reset else 1 = normal reset.
+// one time reset requires a PIN where normal uses a new temp pass sent to user.
+if ($password_update === 2 && !empty($_SESSION['pin'])) {
+    $sql = "SELECT " . implode(",", array(
+            COL_ID, COL_PID, COL_POR_PWD, COL_POR_USER, COL_POR_LOGINUSER, COL_POR_SALT, COL_POR_PWD_STAT, COL_POR_ONETIME)) . " FROM " . TBL_PAT_ACC_ON .
+        " WHERE BINARY " . COL_POR_ONETIME . "= ?";
+    $auth = privQuery($sql, array($_SESSION['forward']));
+    if ($auth !== false) {
+        // remove the token from database
+        sqlStatementNoLog("UPDATE " . TBL_PAT_ACC_ON . " SET " . COL_POR_ONETIME . "=NULL WHERE BINARY " . COL_POR_ONETIME . " = ?", [$auth['portal_onetime']]);
```
Comment (bradymiller) [inline portal/get_patient_info.php]: minor fix for me:
```
sqlStatementNoLog("UPDATE " . TBL_PAT_ACC_ON . " SET " . COL_POR_ONETIME . "=NULL WHERE BINARY " . COL_POR_ONETIME . " = ?", [$auth[COL_POR_ONETIME]]);
```

File: portal/account/account.php (hunk 5 above)
Comment (bradymiller) [reply portal/account/account.php]: for me:
the third comparison above looks good for the call in portal/index.php (just need to ensure the path to this involves auth check; I know it does for when $_SESSION['password_update'] == 2, just need to ensure also does for when it equals 1; going to bed now, but will confirm this tomorrow)

Comment (bradymiller) [reply portal/account/account.php]: ok, just need to test it one more time and then is ready for codebase

**Comment by bradymiller** [comment]
hi @sjpadgett , merging this into master now. I'll get it into rel-502 tomorrow. Awesome stuff btw!

**Comment by bradymiller** [comment]
btw, am in process of working on a API minor update to prevent a pass the hash like vulnerability in the API tokens (it's inspired with how you solved the issue in the portal, but in the opposite direction :) ; only changes are on the backend and should have it ready for PR in a couple days). Goal is to get out a patch this weekend including both the portal stuff and the api.

**Comment by sjpadgett** [comment]
Okay, thanks Brady. I'm sure there will be a few others by then. Meantime I want to test the ICD10 backport. Don't forget I had some other fixes in my PR for this. ICD10 external loads was one and appointments which I think you got.

**Pull request merged**